lib/Sema/CSBindings.cpp

```diff
@@ -538,6 +591,90 @@ void BindingSet::finalize(
       }
     }
 
+    if (TypeVar->getImpl().isKeyPathType()) {
+      auto &ctx = CS.getASTContext();
+
+      auto *keyPathLoc = TypeVar->getImpl().getLocator();
+      auto *keyPath = castToExpr<KeyPathExpr>(keyPathLoc->getAnchor());
+
+      bool isValid;
+      llvm::Optional<KeyPathCapability> capability;
+
+      std::tie(isValid, capability) = CS.inferKeyPathLiteralCapability(TypeVar);
+
+      if (!isValid) {
+        // If key path is invalid we have to drop all the contextual
+        // bindings, none of the could be used unless capability is
+        // known.
+        Bindings.clear();
+
+        // If one of the references in a key path is invalid let's add
+        // a placeholder binding in diagnostic mode to indicate that
+        // the key path cannot be properly resolved.
+        if (CS.shouldAttemptFixes()) {
+          auto rootTy = CS.getKeyPathRootType(keyPath);
+          // If key path is structurally correct and has a resolved root
+          // type, let's promote the fallback type into a binding because
+          // root would have been inferred from explicit type already and
+          // it's benefitial for diagnostics to assign a non-placeholder
+          // type to key path literal to propagate root/value to the context.
+          if (!keyPath->hasSingleInvalidComponent() &&
+              (keyPath->getParsedRoot() ||
+               !CS.getFixedType(rootTy)->isTypeVariableOrMember())) {
+            auto fallback = llvm::find_if(Defaults, [](const auto &entry) {
+              return entry.second->getKind() == ConstraintKind::FallbackType;
+            });
+            assert(fallback != Defaults.end());
+            addBinding(
+                {fallback->first, AllowedBindingKind::Exact, fallback->second});
+          } else {
+            addBinding(PotentialBinding::forHole(
+                TypeVar, CS.getConstraintLocator(
+                             keyPath, ConstraintLocator::FallbackType)));
+          }
+        }
+
+        // No need for fallback if key path is invalid.
+        Defaults.clear();
+        return;
+      }
+
+      // If the key path is sufficiently resolved we can add inferred binding
+      // to the set.
+      if (capability) {
+        SmallSetVector<PotentialBinding, 4> updatedBindings;
+        for (const auto &binding : Bindings) {
+          auto bindingTy = binding.BindingType->lookThroughAllOptionalTypes();
+
+          assert(isKnownKeyPathType(bindingTy) ||
+                 bindingTy->is<FunctionType>());
+
+          // Functions don't have capability so we can simply add them.
+          if (bindingTy->is<FunctionType>())
+            updatedBindings.insert(binding);
```

Comment:
I think in this case we can form a function type that uses root/value type variables and infers parameter flags from the binding, does that make sense, @Jumhyn? If we can make it work then I can remove remaining root/value matching logic from `simplifyKeyPathConstraint`.

Reply:
Actually I think I can approach that from a different angle. I Remembered that there is a hack to eagerly bind key path type if the contextual type is a function type. I think once I remove that everything and replace function result type with key path's value type everything is going to fall into place.

Reply:
Sorry @xedin was out of town last week, but the current solution LGTM!